### PR TITLE
fix: unify names of supported chains

### PR DESCRIPTION
**Pull request by shawnharmsen**
issue: https://github.com/foundry-rs/foundry/issues/1564

Was using references:
https://users.rust-lang.org/t/structopt-enum-like-args/67414/2
https://github.com/TeXitoi/structopt/blob/master/examples/enum_in_args_with_strum.rs

## Solution

Unsure if I should be adding strum
https://github.com/foundry-rs/foundry/blob/c8872a969482b8a863a04effe045774a0b1c494f/cli/Cargo.toml#L73

I was a bit too hurried and didn't try using this:
https://github.com/foundry-rs/foundry/blob/ed4eb2de59ced6c8fc2ce84a101ec9d408b7a552/cli/src/opts/mod.rs#L40-L57

Will try again and update

### CODE REVIEW

File: cli/src/cmd/forge/cache.rs [hunk 1]
```diff
@@ -35,31 +37,44 @@ impl FromStr for ChainOrAll {
...
+    OptimismKovan,
+    Fantom,
+    FantomTestnet,
+    Arbitrum,
+    ArbitrumTestnet,
+    Bsc,
+    BscTestnet,
+    Cronos,
+    CronosTestnet,
+}
```
Comment (mattsse) [inline cli/src/cmd/forge/cache.rs:68]: can we even replace this with the native `ethers-rs` `Chain`?

**Review by mattsse — COMMENTED** [review]
thanks, 
I was wondering since we have a Chain enum on ethers-rs already, can we instead use that?

**Comment by shawnharmsen** [comment]
I'm having a lot of trouble figuring out how to use
`use ethers::prelude::Chain;`

instead of

https://github.com/foundry-rs/foundry/blob/62c728aba7e2a1b8824c7a9d27d4a72a43c75b41/cli/src/cmd/forge/cache.rs#L42-L68


Do you have any pointers/search terms for google fu?

I will keep on at it until I grok it

**Comment by mattsse** [comment]
looks like the https://docs.rs/strum/latest/strum/derive.EnumVariantNames.html macro simply creates a `VARIANTS` constant,

we probably should add this here instead: https://github.com/gakonst/ethers-rs/blob/842f4d260f7f1f7479eaadd7ebf7f98b329d0da5/ethers-core/src/types/chain.rs#L16

so we don't have duplicate Chain types,
we can either add the VARIANTS manually or use strum there

you can test changes to ethers-rs by patching the ethers-rs dependency against a local checkout:
https://github.com/foundry-rs/foundry/blob/b11b776934cce2a0e70ce4879e7a05c9a34ac008/Cargo.toml#L52

**Comment by shawnharmsen** [comment]
awesome will try that

**Comment by shawnharmsen** [comment]
Tests should pass after `ethers-rs `pr gets approved
https://github.com/gakonst/ethers-rs/pull/1249

File: cli/src/cmd/forge/cache.rs [hunk 2]
```diff
@@ -41,25 +42,9 @@ pub struct CleanArgs {
...
-            "avalanche_fuji",
-            "sepolia",
-            "moonbeam",
-            "moonbeam_dev",
-            "moonriver",
-            "optimism",
-            "optimism-kovan"
-    ])]
+        possible_value = "all",
+        possible_values = Chain::VARIANTS
```
Comment (onbjerg) [inline cli/src/cmd/forge/cache.rs:46]: Hmm, AFAIK we already have a `ClapChain` enum somewhere that does this without strum - we should just use that. Also, we should do that everywhere we accept a chain name.

Comment (shawnharmsen) [reply cli/src/cmd/forge/cache.rs:46]: oh whoops, will take a look at using this

**Comment by shawnharmsen** [comment]
I am a bit hard stuck here and would appreciate some guidance

From what I understand the solution should be
- instead of using strum to use `Chain::VARIANTS` from ethers-rs (pr: https://github.com/gakonst/ethers-rs/pull/1249)
- Use the chain enum on the `config` crate

```rust
pub enum Chain {
    /// Contains a known chain
    #[serde(serialize_with = "super::from_str_lowercase::serialize")]
    Named(ethers_core::types::Chain),
    /// Contains the id of a chain
    Id(u64),
}
```

where I am stuck is

If I try to get a `possible_value` from `Chain` from `config/src/chain.rs`

```rust
use foundry_config::{cache, Chain as FoundryConfigChain, Config};

#[derive(Debug, Parser)]
pub struct LsArgs {
    // TODO refactor to dedup shared logic with ClapChain in opts/mod
    #[clap(
        env = "CHAIN",
        default_value = "all",
        possible_value = "all",
        possible_value = FoundryConfigChain::Named(Chain::Mainnet)
    )]
    chains: Vec<ChainOrAll>,
}
```

I get the error below

```sh

error[E0277]: the trait bound `PossibleValue<'_>: std::convert::From<foundry_config::Chain>` is not satisfied
    --> cli/src/cmd/forge/cache.rs:67:26
     |
67   |         possible_value = FoundryConfigChain::Named(Chain::Mainnet)
     |         --------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<foundry_config::Chain>` is not implemented for `PossibleValue<'_>`
     |         |
     |         required by a bound introduced by this call
     |
     = help: the following implementations were found:
               <PossibleValue<'help> as std::convert::From<&'help &'help str>>
               <PossibleValue<'help> as std::convert::From<&'help str>>
     = note: required because of the requirements on the impl of `Into<PossibleValue<'_>>` for `foundry_config::Chain`
note: required by a bound in `Arg::<'help>::possible_value`
    --> /Users/m1/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-3.1.12/src/build/arg.rs:1720:12
     |
1720 |         T: Into<PossibleValue<'help>>,
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Arg::<'help>::possible_value`

For more information about this error, try `rustc --explain E0277`.
warning: `foundry-cli` (bin "forge") generated 2 warnings
error: could not compile `foundry-cli` due to previous error; 2 warnings emitted
warning: build failed, waiting for other jobs to finish...
warning: `foundry-cli` (bin "cast") generated 2 warnings (2 duplicates)
error: build failed
````

**Comment by shawnharmsen** [comment]
I am a bit hard stuck here and would appreciate some guidance

From what I understand the solution should be
- instead of using strum to use `Chain::VARIANTS` from ethers-rs (pr: https://github.com/gakonst/ethers-rs/pull/1249)
- Use the chain enum on the `config` crate

```rust
pub enum Chain {
    /// Contains a known chain
    #[serde(serialize_with = "super::from_str_lowercase::serialize")]
    Named(ethers_core::types::Chain),
    /// Contains the id of a chain
    Id(u64),
}
```

where I am stuck is

If I try to get a `possible_value` from `Chain` from `config/src/chain.rs`

```rust
use foundry_config::{cache, Chain as FoundryConfigChain, Config};

#[derive(Debug, Parser)]
pub struct LsArgs {
    // TODO refactor to dedup shared logic with ClapChain in opts/mod
    #[clap(
        env = "CHAIN",
        default_value = "all",
        possible_value = "all",
        possible_value = FoundryConfigChain::Named(Chain::Mainnet)
    )]
    chains: Vec<ChainOrAll>,
}
```

I get the error below

```sh

error[E0277]: the trait bound `PossibleValue<'_>: std::convert::From<foundry_config::Chain>` is not satisfied
    --> cli/src/cmd/forge/cache.rs:67:26
     |
67   |         possible_value = FoundryConfigChain::Named(Chain::Mainnet)
     |         --------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<foundry_config::Chain>` is not implemented for `PossibleValue<'_>`
     |         |
     |         required by a bound introduced by this call
     |
     = help: the following implementations were found:
               <PossibleValue<'help> as std::convert::From<&'help &'help str>>
               <PossibleValue<'help> as std::convert::From<&'help str>>
     = note: required because of the requirements on the impl of `Into<PossibleValue<'_>>` for `foundry_config::Chain`
note: required by a bound in `Arg::<'help>::possible_value`
    --> /Users/m1/.cargo/registry/src/github.com-1ecc6299db9ec823/clap-3.1.12/src/build/arg.rs:1720:12
     |
1720 |         T: Into<PossibleValue<'help>>,
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Arg::<'help>::possible_value`

For more information about this error, try `rustc --explain E0277`.
warning: `foundry-cli` (bin "forge") generated 2 warnings
error: could not compile `foundry-cli` due to previous error; 2 warnings emitted
warning: build failed, waiting for other jobs to finish...
warning: `foundry-cli` (bin "cast") generated 2 warnings (2 duplicates)
error: build failed
````

**Comment by onbjerg** [comment]
@shawnharmsen I think using `ClapChains` might be preferable here because you need to derive `ArgEnum` for the enum like so:

```rs
#[derive(ArgEnum)]
pub enum MyEnum {
  OptionOne,
  OptionTwo
}
```

Doing that it just automagically kind of works. See the `CoverageReportKind` enum I have in the coverage PR (in `cli/src/cmd/forge/coverage.rs`)

**Comment by shawnharmsen** [comment]
Apologies but ran into another problem

Latest commit I see how argEnum works

```rust
pub struct LsArgs {
\\ ...
#[clap(long, arg_enum, default_value = "all", help = "Name of chain")]
    chains2: ChainOptions,
}

#[derive(Debug, Clone, ArgEnum)]
pub enum ChainOptions {
    All,
    Mainnet,
    Goerli,
}
```

and the `--help` shows the new --chains2 with possible values

```sh
foundry git:(feat-unify-names-supported-chains) target/debug/forge cache ls -h
forge-cache-ls 
Shows cached data from ~/.foundry.

USAGE:
    forge cache ls [OPTIONS] [CHAINS]...

ARGS:
    <CHAINS>...    [env: CHAIN=] [default: all] [possible values: all, mainnet, ropsten, rinkeby, goerli, kovan, xdai, polygon, polygon-mumbai, avalanche, avalanche-fuji, sepolia,
                   moonbeam, moonbeam-dev, moonriver, optimism, optimism-kovan, fantom, fantom-testnet, arbitrum, arbitrum-testnet, bsc, bsc-testnet, cronos]

OPTIONS:
        --chains2 <CHAINS2>    Name of chain [default: all] [possible values: all, mainnet, goerli]
    -h, --help                 Print help information

```


When I try to use `argEnum` on `enum ChainOrAll` I get `ArgEnum only supports non-unit variants`
```sh
 foundry git:(feat-unify-names-supported-chains) ✗ cargo build
   Compiling foundry-cli v0.2.0 (/Users/m1/CLionProjects/foundry/cli)
error: `#[derive(ArgEnum)]` only supports non-unit variants, unless they are skipped
  --> cli/src/cmd/forge/cache.rs:20:5
   |
20 |     Chain(Chain),
   |     ^^^^^

```

**Comment by shawnharmsen** [comment]
I've opened a new pr because I've been getting build errors that I think stem from cargo.lock

Will close this one and use this: https://github.com/foundry-rs/foundry/pull/1636

**Pull request closed without merging**